pkg/sysconfwatch/sysconfwatch.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright © 2022 Microshift Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sysconfwatch
+
+import (
+	"context"
+	"math"
+	"os"
+	"time"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
+)
+
+const sysConfigCheckInterval = time.Second * 5
+const sysConfigAllowedTimeDrift = time.Second * 10
+
+type SysConfWatchController struct {
+	NodeIP  string
+	timerFd int
+}
+
+func NewSysConfWatchController(cfg *config.MicroshiftConfig) *SysConfWatchController {
+	// Create a realtime clock timer with asynchronous read support
+	fd, err := unix.TimerfdCreate(unix.CLOCK_REALTIME, unix.TFD_CLOEXEC|unix.TFD_NONBLOCK)
+	if err != nil {
+		klog.Fatalf("failed to create a realtime clock timer %v", err)
+	}
+
+	// Set the time interval into distant future
+	var ptime = &unix.ItimerSpec{
+		Interval: unix.Timespec{Sec: math.MaxInt64, Nsec: 0},
+		Value:    unix.Timespec{Sec: 0, Nsec: 0},
+	}
+	// Start the timer with cancelation if the clock is reset elsewhere
+	err = unix.TimerfdSettime(fd, unix.TFD_TIMER_ABSTIME|unix.TFD_TIMER_CANCEL_ON_SET, ptime, nil)
+	if err != nil {
+		klog.Fatalf("failed to start a realtime clock timer %v", err)
+	}
+
+	return &SysConfWatchController{
+		NodeIP:  cfg.NodeIP,
+		timerFd: fd,
+	}
+}
+
+func (s *SysConfWatchController) Name() string { return "sysconfwatch-controller" }
+func (s *SysConfWatchController) Dependencies() []string {
+	return []string{}
+}
+
+func sysMonTimeDiff() int64 {
+	var stm unix.Timespec
+	var mtm unix.Timespec
+
+	// System-wide clock that measures real (i.e. wall-clock) time
+	// This clock is affected by discontinuous jumps in the system time (e.g. if the system administrator manually changes the clock)
+	// and by the incremental adjustments performed by adjtime and NTP
+	unix.ClockGettime(unix.CLOCK_REALTIME, &stm)
+	// Clock that cannot be set and represents monotonic time since some unspecified starting point
+	// It provides access to a raw hardware-based time that is not subject to NTP adjustments
+	// or the incremental adjustments performed by adjtime
+	unix.ClockGettime(unix.CLOCK_MONOTONIC_RAW, &mtm)
+
+	return stm.Sec - mtm.Sec
```

Comment:
The difference between monotonic (continuously increasing for the process lifetime) and wall (susceptible to adjustments) should remain the same as long as wall time does not change. 

What you are suggesting will also work, but it is a bit more complicated check. Are you concerned that we are not comparing "apples-to-apples"? I can change it if you think it's a better way to code this.

Reply:
Yes, I'm worried that the 2 clocks are different enough in nature that relying on them to consistently have the same value would lead to a lot of false-positives and restarts. Watching for a difference in the way they each measure their own notion of elapsed time shouldn't have that effect.